Correctly handle object cycles in munchify/unmunchify

airbornemint opened the pull request.

vmalloc commented on test_munch.py:
```diff
@@ -348,6 +348,27 @@ def test_munchify_default_factory():
+    
+    d = unmunchify(x)
+    assert d["id"] == "x"
+    assert d["y"]["id"] == "y"
```
Please add checks using `is`, making sure objects aren't only preserved by value, but by identity as well

vmalloc commented:
Thanks! I have a small request regarding the test(s), and then I'll merge your change.

airbornemint commented:
Good point!

I also added more tests, which resulted in me discovering that tuples were not handled correctly in some cases, which is now fixed.

The bug was that if the first object encountered in a cycle was a tuple, it would not be handled correctly. For example, although the object graph of dictX → tupleA → dictX would correctly give you munchX → tupleC → munchX, the object graph dictX → tupleA → dictY → tupleA would incorrectly give you munchX → tupleC → dictY → tupleD → dictY. Now it correctly gives munchX → tupleC → dictY → tupleC.

vmalloc commented on munch/__init__.py:
```diff
@@ -397,13 +397,41 @@ def munchify(x, factory=Munch):
 
         nb. As dicts are not hashable, they cannot be nested in sets/frozensets.
     """
+    return cycle_munchify(x, factory, dict())
```
On second glance I'm not sure we should prefix everything with `cycle_`... Given this will become the default behavior, we're better off just fixing munchify...

airbornemint replied:
To clarify, I already did change `munchify`/`unmunchify` to just call the `cycle_` variants, so I did "just fix" them already. However, you can't implement this code without having `munchify` call through to `cycle_munchify`, because you need a place to create the `seen` dict.

Addressing your other comments, I'd be happy to improve naming of the 6 helper functions, but (due to the way they recursively call each other), I don't think the code would improve in clarity or correctness if we try to eliminate them. That said, depending on how you stylistically feel about functions inside functions, I can make all of this less sprawly by using them. 

I'll drop a commit to that effect in a few minutes so you can evaluate what that would actually look like.

vmalloc replied:
We can leave the `cycle_` prefix, but please make these module-private (see other comment I made)

airbornemint replied:
Have a look at the latest changes I pushed; the helper functions for breaking cycles are not at module level anymore, so the issue of being module-private is moot.

vmalloc replied:
Gotcha

vmalloc commented on munch/__init__.py:
```diff
@@ -397,13 +397,41 @@ def munchify(x, factory=Munch):
+    cycle_postmunchify(m, x, factory, seen)
+    return m
+
+def cycle_premunchify(x, factory, seen):
```
Assuming these are not intended for outside use, please prefix with `_` to prevent accidental importing

vmalloc commented on munch/__init__.py:
```diff
@@ -397,13 +397,41 @@ def munchify(x, factory=Munch):
+    elif isinstance(x, list):
+        m.extend(cycle_munchify(v, factory, seen) for v in x)
+    elif isinstance(x, tuple):
+        for (mv, v) in zip(m, x):
```
please give better names to these... Too many variables spread around with single letters is very confusing to follow

vmalloc replied:
This comment is still relevant

airbornemint replied:
Should be better / good now

vmalloc commented:
LGTM. Thanks for your contribution!

airbornemint commented:
And thanks for your feedback! 🙂
